### PR TITLE
add  annotation on cluster for cert request expire timestamp

### PR DESCRIPTION
**Pull request by alinadonisa**
## Synopsis

The purpose of this PR is to provide a mechanism to manually rotate certificates. First step will be to signal the user the expiration date of the node certificate, this way the user can acknowledge that he/she needs to manually replace the existing certificate and restart the cluster afterwards.

We are taking in consideration 2 types of restart:

- **FullCluster** restart that will be use when CA certificate is replaced
- **Rolling** restart used in the other cases

For future development we are taking in consideration also to add integration with certmanager.

## Work completed this week

- I added implementation for Rolling restart and FullCluster restart
- I fixed error from sync version checker and request cert action 
- I worked on refactory  the code for version checker 

## Work pending:

-  Restart runs only once now, we should think of a method to run it once and not interfere with the operator workflow.

### CODE REVIEW

File: pkg/resource/cluster.go [hunk 1]
```diff
@@ -204,6 +205,12 @@ func (cluster Cluster) SetAnnotationContainerImage(containerimage string) {
 	}
 	cluster.cr.Annotations[CrdbContainerImageAnnotation] = containerimage
 }
+func (cluster Cluster) SetAnnotationCAExpiration(version string) {
```
Comment (chrislovecnm) [inline pkg/resource/cluster.go:208]: Need godocs please

File: pkg/resource/cluster.go [hunk 2]
```diff
@@ -38,6 +38,7 @@ const (
 	CrdbContainerImageAnnotation = "crdb.io/containerimage"
 	CrdbVersionAnnotation        = "crdb.io/version"
 	CrdbHistoryAnnotation        = "crdb.io/history"
+	CrdbCAExpirationAnnotation   = "crdb.io/certexpiration"
```
Comment (chrislovecnm) [inline pkg/resource/cluster.go:41]: One thing I noticed is that crdb.io does not exist. @keith-mcclellan, which domain do we want on these?

**Review by chrislovecnm — COMMENTED** [review]
Just a couple of comments and please check on CI

File: pkg/resource/cluster.go (hunk 2 above)
Comment (keith-mcclellan) [reply pkg/resource/cluster.go]: why do we have a url here at all @chrislovecnm ?

File: pkg/actor/cluster_restart.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+	config *rest.Config
+}
+
+//GetActionType returns api.VersionCheckerAction action used to set the cluster status errors
+func (v *clusterRestart) GetActionType() api.ActionType {
+	return api.ClusterRestartAction
+}
+
+func (v *clusterRestart) Handles(conds []api.ClusterCondition) bool {
```
Comment (chrislovecnm) [inline pkg/actor/cluster_restart.go:53]: Godocs please

File: pkg/resource/cluster.go [hunk 4]
```diff
@@ -204,6 +205,12 @@ func (cluster Cluster) SetAnnotationContainerImage(containerimage string) {
 	}
 	cluster.cr.Annotations[CrdbContainerImageAnnotation] = containerimage
 }
+func (cluster Cluster) SetAnnotationCertExpiration(certExpiration string) {
```
Comment (chrislovecnm) [inline pkg/resource/cluster.go:208]: Godocs

**Review by keith-mcclellan — CHANGES_REQUESTED** [review]
In general I think this does what we want - we can merge once you resolve Chris's comments.  In the future, I'd have preferred that a feature like ClusterRestart be done in it's own PR because there are other places to use that feature than just rotating the certs and we want to make sure the implementation is general enough.  It is much easier to review one feature at a time than two together.  Thanks!!!

File: pkg/resource/cluster.go (hunk 2 above)
Comment (alinadonisa) [reply pkg/resource/cluster.go]: @keith-mcclellan if the prefix is omitted, the annotation Key is presumed to be private to the user. Automated system components (e.g. kube-scheduler, kube-controller-manager, kube-apiserver, kubectl, or other third-party automation) which add annotations to end-user objects must specify a prefix. Please see k8s documentation for annotations: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/

**Comment by chrislovecnm** [comment]
@alinadonisa we need some godocs updates on this code please, and then we can merge

File: Makefile [hunk 5]
```diff
@@ -18,13 +18,13 @@
 #
 
 # values used in workspace-status.sh
-DOCKER_REGISTRY?=cockroachdb
+DOCKER_REGISTRY?=us.gcr.io/chris-love-operator-playground
```
Comment (chrislovecnm) [inline Makefile:21]: Need to roll this back please

File: manifests/operator.yaml [hunk 6]
```diff
@@ -257,16 +257,16 @@ spec:
       serviceAccountName: cockroach-operator-sa
       containers:
         - name: cockroach-operator
-          image: cockroachdb/cockroach-operator:v1.6.12-rc.2
+          image: us.gcr.io/chris-love-operator-playground/cockroachdb-operator:v1.9.17
```
Comment (chrislovecnm) [inline manifests/operator.yaml:260]: Back out changes please

File: pkg/actor/cluster_restart.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+	config *rest.Config
+}
+
+//GetActionType returns api.ClusterRestartAction action used to set the cluster status errors
+func (r *clusterRestart) GetActionType() api.ActionType {
+	return api.ClusterRestartAction
+}
+
+func (r *clusterRestart) Handles(conds []api.ClusterCondition) bool {
```
Comment (chrislovecnm) [inline pkg/actor/cluster_restart.go:64]: GoDocs please

File: pkg/actor/cluster_restart.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+	return utilfeature.DefaultMutableFeatureGate.Enabled(features.ClusterRestart) && condition.True(api.InitializedCondition, conds) || condition.False(api.InitializedCondition, conds)
+}
+
+func (r *clusterRestart) Act(ctx context.Context, cluster *resource.Cluster) error {
+	log := r.log.WithValues("CrdbCluster", cluster.ObjectKey())
+	log.V(int(zapcore.DebugLevel)).Info("starting cluster restart action")
```
Comment (chrislovecnm) [inline pkg/actor/cluster_restart.go:70]: Please use the log package I wrote

File: pkg/actor/cluster_restart.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+	if cluster.Spec().RestartType == api.RollingRestart {
+		log.V(int(zapcore.DebugLevel)).Info("initiating rolling restart action")
+		// TODO statefulSetIsUpdating is not quite working as expected.
+		// I had to check status.  We should look at the update code in partition update to address this
+		if statefulSetIsUpdating(statefulSet) {
+			return NotReadyErr{Err: errors.New("restart statefulset is updating, waiting for the update to finish")}
+		}
+		status := &statefulSet.Status
+		if status.CurrentReplicas == 0 || status.CurrentReplicas < status.Replicas {
+			log.Info("restart statefulset does not have all replicas up")
```
Comment (chrislovecnm) [inline pkg/actor/cluster_restart.go:98]: Use logging levels please

File: pkg/actor/cluster_restart.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+			return NotReadyErr{Err: errors.New("restart cluster statefulset does not have all replicas up")}
+		}
+		timeNow := metav1.Now()
+		// stsName := statefulSet.Name
```
Comment (chrislovecnm) [inline pkg/actor/cluster_restart.go:102]: Remove dead code

File: pkg/actor/cluster_restart.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+	}
+	return update.RollingRestart(ctx, updateRoach, k8sCluster, r.log)
+}
+
+//ScaleSts updates the replicas of the statefullset to the value from parameters
+//and waits until the statefullset has currentreplicas equal with the desired replicas
+//We will use this in FullRestart logic, by setting replicas to 0 and
+//afterwords setting to the original number of nodes, this way the cluster
+//will load the new CA certs
+func (r *clusterRestart) ScaleSts(ctx context.Context, sts *appsv1.StatefulSet, l logr.Logger, clientset *kubernetes.Clientset, replicas int32) error {
```
Comment (chrislovecnm) [inline pkg/actor/cluster_restart.go:176]: Does this need to be public func?

**Review by chrislovecnm — COMMENTED** [review]
Nitpicks mostly and we need a test

File: apis/v1alpha1/cluster_types.go [hunk 12]
```diff
@@ -97,6 +97,13 @@ type CrdbClusterSpec struct {
 	// Default: ""
 	// +optional
 	CockroachDBVersion string `json:"cockroachDBVersion,omitempty"`
+	// (Optional) RestartType if set with value FullCluster restarts the entire cluster.
+	// FullCluster option is used when the CA cert is replaced and we need to restart all nodes at once.
+	// If the value is set to Rollback it will restart one node at a time.
```
Comment (udnay) [inline apis/v1alpha1/cluster_types.go:102]: /typo? `Rollback` should be `Rolling`?

File: config/crd/bases/crdb.cockroachlabs.com_crdbclusters.yaml [hunk 13]
```diff
@@ -260,6 +260,9 @@ spec:
                     description: 'Requests describes the minimum amount of compute resources required. If Requests is omitted for a container, it defaults to Limits if that is explicitly specified, otherwise to an implementation-defined value. More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/'
                     type: object
                 type: object
+              restartType:
+                description: '(Optional) RestartType if set with value FullCluster restarts the entire cluster. FullCluster option is used when the CA cert is replaced and we need to restart all nodes at once. If the value is set to Rollback it will restart one node at a time. Other values are invalid. Default: ""'
```
Comment (udnay) [inline config/crd/bases/crdb.cockroachlabs.com_crdbclusters.yaml:264]: Same comment as above

File: pkg/resource/cluster.go (hunk 2 above)
Comment (keith-mcclellan) [reply pkg/resource/cluster.go]: got it thanks

**Review by keith-mcclellan — APPROVED** [review]
LGTM

**Pull request merged**